Mark infinitely retrying jobs as FAILED

flodnv opened the pull request:
### Details

Some jobs can retry forever (if no one ever calls FinishJob / FailJob, but keeps calling GetJob(s)). This fixes that by adding a maximum number of 10 retries. 

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/179455

### Tests

Automated

iwiznia commented on plugins/Jobs.cpp:
```diff
@@ -787,16 +787,29 @@ void BedrockJobsCommand::process(SQLite& db) {
+                    if (!db.writeIdempotent(failQuery)) {
+                        STHROW("502 Update failed");
+                    }
+                    // TODO: Remove job from returned jobs...?
```
I think we need to remove it, no? Otherwise when it runs and tries to finish/fail it, it will fail?

tylerkaraszewski replied:
I think the same as @iwiznia , we shouldn't return it or it'll run again and probably throw some error when we see it's already failed.

iwiznia commented on plugins/Jobs.cpp:
```diff
@@ -787,16 +787,29 @@ void BedrockJobsCommand::process(SQLite& db) {
+                                     "SET state = 'RUNQUEUED', "
+                                         "lastRun = " + SQ(currentTime) + ", " +
+                                         "nextRun = " + nextRunDateTime + ", " +
+                                         "data = JSON_SET(data, '$.retryAfterCount', COALESCE(JSON_EXTRACT(data, '$.retryAfterCount'), 0) + 1" + // Set this so we don't retry infinitely (see above)
```
You already have the data, so you could directly set it, instead of doing the coalesce and + 1 in sql? NAB

iwiznia commented on test/tests/jobs/InifiniteRetryAfterTest.cpp:
```diff
@@ -0,0 +1,47 @@
+        // Create a job
+        SData createJob("CreateJob");
+        createJob["name"] = "infinite-job";
+        createJob["retryAfter"] = "+1 SECOND";
```
```suggestion
        createJob["retryAfter"] = "QUEUED +1 SECOND";
```
?

flodnv replied:
No? 

![image](https://user-images.githubusercontent.com/4164921/137346655-dd580e4b-e92f-45bf-9eaa-58a5724f6f8d.png)

bondydaa commented on plugins/Jobs.cpp:
```diff
@@ -787,16 +787,29 @@ void BedrockJobsCommand::process(SQLite& db) {
             for (auto job : retriableJobs) {
                 SINFO("Updating job with retryAfter " << job["jobID"]);
+                STable jobData = SParseJSONObject(job["data"]);
+                if (SToInt(jobData["retryAfterCount"]) >= 10) {
```
I was wondering if we needed to check if `jobData["retryAfterCount"]` exists but looking at what `SToInt` does: https://github.com/Expensify/Bedrock/blob/9b8fe5493e3ae6787830a65c6930cc631ec9164b/libstuff/libstuff.cpp#L2739-L2741

we're fine right? This is because the `val.c_str()` will return a c string that is null terminated (`\0`) https://www.cplusplus.com/reference/string/string/c_str/ and `atoi` will return 0 if the passed c string has only whitespace characters or a non valid integer value. 

so if `jobData["retryAfterCount"]` isn't set for a job then the c string returned by `val.c_string()` will just be the null terminated value and `atoi` will return 0 for that. 

Just rubber ducking this out to make sure I'm following it properly.

bondydaa reviewed (dismissed):
I think we're good here, I'm very unsure about if we should or shouldn't be returning the job where you've called it out but should be an easy fix if we need to not return it so I'm fine as is.

flodnv commented:
Updated:
- Fixed unrelated test
- Remove job from returned jobs... I am not a fan of this solution, the other option is a small refactor (because we store strings inside the `jobList`...
- Last night after pushing this PR I realized it had a massive flaw: it would've broken all repeating jobs with a retryAfter 😅 I fixed that in the last commit: In FinishJob, remove the retryAfterCount value.
